Implement Servo() on MicoArm

sjavdani opened the pull request:
Implemented joint velocity commands through the Servo function for the mico arm. Also added a watchdog that sends zero velocity after 0.25 seconds of no commands sent to the arm.

Not included is the code to activate all the controllers - I'll add that directly to adapy, since it uses the controller interface included there.

mkoval commented on src/prpy/base/mico.py:
```diff
@@ -99,4 +109,11 @@ def Servo(self, velocities):
+            self.servo_watchdog.reset()
+
+
+    def SendVelocitiesToMico(self, velocities):
```
I'd prefer to put this in `Servo`, since this code has a subset of that functionality: it doesn't work in simulation. Is there a strong reason to keep it as a separate function?

sjavdani replied:
This is so the watchdog can call something without going through Servo. If I had it in servo, and the watchdog called Servo, then it would reset itself

mkoval replied:
Good point! :+1:

mkoval commented on src/prpy/util.py:
```diff
@@ -497,6 +497,21 @@ def stop(self):
+        
+    def stop(self):
+        self.timer.cancel()
+
```
Please add docstrings to this class.

mkoval commented on src/prpy/util.py:
```diff
@@ -497,6 +497,21 @@ def stop(self):
+    def reset(self):
+        self.timer.cancel()
+        self.timer = threading.Timer(self.timeout_duration, self.handler)
+        self.timer.start()
```
I am a bit concerned about creating a new `Timer` every time we send a velocity command. My hunch is that every `Timer` starts a new thread. This will create and destroy a huge number of threads. There may be some gotchas about this, e.g. thread exhaustion, performance implications, but I am not positive. Maybe @cdellin or @psigen is more informed?

If this indeed an issue, then it is relatively straightforward to use one long-running thread per `Watchdog` by setting the expiration timestamp as a member variable.

mkoval commented on src/prpy/base/mico.py:
```diff
@@ -99,4 +109,11 @@ def Servo(self, velocities):
+
+    def SendVelocitiesToMico(self, velocities):
+        for velocity_publisher,velocity in zip(self.velocity_publishers, velocities):
+            velocity_publisher.publish(velocity)
```
You should introduce a `Lock` around this loop. This function is called from the user's thread(s) and the watchdog thread. I am not sure if `publish` is threadsafe. Even if it is, bad things will happen if `publish` commands are interleaved between two threads.

mkoval commented:
The high-level implementation looks good. I posted a few comments related to threading on the commits.

sjavdani commented:
Ok, added lock, added docstring and modified the use of timer. Comment about servo vs. set velocity inline. Now, it will reset it's internal timer until the handler is called once, at which time the thread will exit. if you call reset again, and the thread has exited, it will spawn a new thread.
